### PR TITLE
Cleanbot: Ability to restrict targeting to map areas

### PR DESCRIPTION
**Pull request by Oyu07**
## What Does This PR Do

You are able to lock a cleanbot to an area so it does not wonder off to 7th dimension

Adds glass to things cleanbot can clean

Makes cleanbot icon turn YELLOW
I am adapting it from medibot, making YELLOW = Target / Movement restriction 

Makes cleanbots use overlays

## Why It's Good For The Game,

Departmental cleanbots are real. When you want a cleanbot in your department, like bar or medbay to clean while janitor is fighting their inner demons, the cleanbot usually wonders off because it will follow the trail. Now it wont

Overlays are one step closer to sprite cleanup

## Images of changes

 N/A
Might add later

## Testing

Spawned stuff to clean, observe what it cleans and where it cleans

## Changelog
:cl:
tweak: Cleanbots can be restricted to an area
tweak: Cleanbots now clean glass
/:cl:

### CODE REVIEW

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 1]
```diff
@@ -41,7 +42,9 @@
 		/obj/effect/decal/cleanable/flour,
 		/obj/effect/decal/cleanable/ash,
 		/obj/effect/decal/cleanable/greenglow,
-		/obj/effect/decal/cleanable/dirt
+		/obj/effect/decal/cleanable/dirt,
+		/obj/effect/decal/cleanable/glass
+
 	)
 	var/static/list/clean_blood = list(
```
Comment (1080pCat) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:49]:
```suggestion
		/obj/effect/decal/cleanable/dirt,
		/obj/effect/decal/cleanable/glass
	)

	var/static/list/clean_blood = list(
```

**Comment by Oyu07** [comment]
I will do a proper sprite overlay / cleanup after this pr. Going to touch the floorbot sprite aswell

**Review by 1080pCat — APPROVED** [review]

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 2]
```diff
@@ -61,7 +63,8 @@
 
 /mob/living/simple_animal/bot/cleanbot/turn_on()
 	..()
-	icon_state = "cleanbot[on]"
+	icon_state = "cleanbot1[area_locked ? "r" : null]"
```
Comment (Henri215) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:66]: r? Not better to name it "_locked" instead?

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 3]
```diff
@@ -61,7 +63,8 @@
...
+
 
 /mob/living/simple_animal/bot/cleanbot/turn_off()
 	..()
```
Comment (Henri215) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:70]: About the line below, you could change it to "cleanbot0" following the logic you used on turn_on()

**Review by Henri215 — COMMENTED** [review]
image
 touched here by accident

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 2 above)
Comment (Oyu07) [reply code/modules/mob/living/simple_animal/bot/cleanbot.dm:66]: Mostly laziness. I am going to turn power icons (green and yellow) to overlays after this

**Comment by Oyu07** [comment]
> image touched here by accident

I dunno what changed, its the exact same. didnt even open that sprite. Should be fine

**Comment by Oyu07** [comment]
Known sprite / icon issue. 
Fixing Tomorrow

**Comment by Danchi299** [comment]
I think all bots should have "statonary mode"
but instead of not moving, they simply return to position where "statonary mode" was activated when they have nothing left to do

**Comment by Oyu07** [comment]
> I think all bots should have "statonary mode" but instead of not moving, they simply return to position where "statonary mode" was activated when they have nothing left to do

I get what you are saying, perhaps it would be somewhat beneficial for cleanbot. This implementation is mostly for medical, bar and maybe brig and cargo to use it wtithout the need for janitor for small stuff. It could be made like that so it comes back, but this could possibly lead to situation where it just follows a blood trail all the way to mining for example, which is the thing we are trying to avoid

As for "all bots should have this", current usecase and general accepted settings for medibot (put one in genetics and medical, stationary so people can use it, and its always guaranteed to be there)

For floorbots, I do not know, the idea of floorbots patrolling funny to me, what are they gonna find, a random damaged tile in the hallway? perhaps it could also have a call sysyem by the engineers?

**Comment by Oyu07** [comment]
Sprites converted to overlays + actually tested

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 4]
```diff
@@ -59,19 +61,26 @@
...
-/mob/living/simple_animal/bot/cleanbot/turn_on()
-	..()
-	icon_state = "cleanbot[on]"
+/mob/living/simple_animal/bot/cleanbot/update_icon_state()
+	return
 
-/mob/living/simple_animal/bot/cleanbot/turn_off()
-	..()
-	icon_state = "cleanbot[on]"
+/mob/living/simple_animal/bot/cleanbot/update_overlays() // Using ``clean_`` before remaking other bot. each bot sprite is offset, overlays would get mixed up
```
Comment (Henri215) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:67]:
```suggestion
/mob/living/simple_animal/bot/cleanbot/update_overlays()
```

**Review by Henri215 — APPROVED** [review]
Lewcing good to me

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 5]
```diff
@@ -59,19 +61,26 @@
...
+/mob/living/simple_animal/bot/cleanbot/update_overlays()
+	. = ..()
+	if(!on)
+		. += "clean_off"
+		return
+	if(mode != BOT_CLEANING)
+		. += "clean_[area_locked ? "restrict" : "on"]"
+	else
+		. += "clean_brush"
+		. += "clean[area_locked ? "_restrict" : ""]_work"
```
Comment (Contrabang) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:76]: For consistency
```suggestion
		. += "clean_[area_locked ? "restrict_" : ""]work"
```

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 6]
```diff
@@ -100,8 +109,14 @@
...
-	if(is_type_in_typecache(D, clean_dirt) || blood && is_type_in_typecache(D, clean_blood))
-		return D
+	if(!(is_type_in_typecache(D, clean_dirt) || blood && is_type_in_typecache(D, clean_blood)))
+		return FALSE
+	if(area_locked)
+		var/area/target_area = get_area(D)
+		if(target_area == area_locked)
+			return D
+		return FALSE
+	return D
```
Comment (Contrabang) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:119]:
```suggestion
	if(!area_locked)
		return F
	var/area/target_area = get_area(D)
	if(target_area == area_locked)
		return D
	return FALSE
```

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm [hunk 7]
```diff
@@ -157,11 +172,18 @@
 
 	oldloc = loc
 
+/mob/living/simple_animal/bot/cleanbot/proc/assign_area()
+	if(area_locked)
+		area_locked = null
+	else
+		area_locked = get_area(loc)
```
Comment (Contrabang) [inline code/modules/mob/living/simple_animal/bot/cleanbot.dm:179]: Just for readability
```suggestion
		area_locked = get_area(src)
```

**Review by Contrabang — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/bot/cleanbot.dm (hunk 6 above)
Comment (Oyu07) [reply code/modules/mob/living/simple_animal/bot/cleanbot.dm:119]: ok I thought I changed this.

**Review by Contrabang — APPROVED** [review]
lgtm

**Review by Contrabang — APPROVED** [review]

**Pull request merged**